CRM/Activity/Tokens.php

```diff
@@ -98,20 +138,122 @@ public function alterActionScheduleQuery(\Civi\ActionSchedule\Event\MailingQuery
     }
   }
 
+  /**
+   * @inheritDoc
+   */
+  public function prefetch(\Civi\Token\Event\TokenValueEvent $e) {
+    if (!empty($e->getTokenProcessor()->context['actionMapping'])) {
+      // Scheduled reminders provide results objects so don't prefetch
+      return NULL;
+    }
+
+    // Get the activities
+    foreach ($e->getRows() as $row) {
+      $activityIds[] = $row->context['activityId'];
+    }
+
+    $activities = civicrm_api3('Activity', 'get', array(
+      'id' => array('IN' => $activityIds),
+    ));
+    $prefetch['activity'] = $activities['values'];
+
+    // If we need ActivityContacts, load them
+    $messageTokens = $e->getTokenProcessor()->getMessageTokens();
+    $needContacts = FALSE;
+    foreach ($messageTokens[$this->entity] as $token) {
+      if (preg_match('/^source|target|assignee/', $token)) {
+        $needContacts = TRUE;
+        break;
+      }
+    }
+    if ($needContacts) {
+      $result = civicrm_api3('ActivityContact', 'get', [
+        'sequential' => 1,
+        'activity_id' => array('IN' => $activityIds),
+      ]);
+      $contactIds = [];
+      $types = [ '1' => 'assignee', '2' => 'source', '3' => 'target'];
+      foreach ($result['values'] as $ac) {
+        if ($ac['record_type_id'] == 2) {
+          $prefetch['activityContact'][$ac['activity_id']][$types[$ac['record_type_id']]] = $ac['contact_id'];
+        }
+        else {
+          $prefetch['activityContact'][$ac['activity_id']][$types[$ac['record_type_id']]][] = $ac['contact_id'];
+        }
+        $contactIds[$ac['contact_id']] = 1;
+      }
+      $result = civicrm_api3('Contact', 'get', [
+        'id' => array('IN' => array_keys($contactIds)),
+      ]);
+      $prefetch['contact'] = $result['values'];
+    }
+    return $prefetch;
+  }
+
   /**
    * @inheritDoc
    */
   public function evaluateToken(\Civi\Token\TokenRow $row, $entity, $field, $prefetch = NULL) {
-    $actionSearchResult = $row->context['actionSearchResult'];
+    // maps token name to api field
+    $mapping = array(
+      'activity_id' => 'id',
+    );
+
+    if (!empty($row->context['actionSearchResult'])) {
+      // For scheduled reminders
+      $activity = $row->context['actionSearchResult'];
+    }
+    else {
+      $activity = (object) $prefetch['activity'][$row->context['activityId']];
+    }
```

Comment:
I'm not entirely certain where to to put this comment; maybe this works...

There's a sort of elegance in saying that `$activity` is the same thing whether it comes from the `context['actionSearchResult']` or from `$prefetch`. Intuitively, these both provide a copy of `civicrm_activity.*`, so that feels well and good.

The problem is that they're not really interchangeable. For example:

* `actionSearchResult` includes `activity_type` (the label from the OV); however, `$prefetch` comes from a default `Activity.get` API call, so it doesn't.
* `actionSearchResult` includes `case_id`; however, `$prefetch` comes from a default `Activity.get` API call, so it doesn't.
* `$prefetch` includes information about related contacts (given the `activityId`); however, `prefetch()` short-circuits in the presence of `actionSearchResult`. So the contacts are consistently prefetched.
* There's an expectation in our community that the schema for `Activity.get` API and for `civicrm_activity` are allowed to drift apart. `actionSearchResult` implicitly tracks one while `$prefetch` implicitly tracks the other.

IMHO, it would get confusing (over time) to maintain with two different functions for loading the data (`alterActionScheduleQuery()` and `prefetch()`).

Of course, `alterActionScheduleQuery()` is in there for a reason -- reducing the number of queries by tapping into the main query. We don't want to issue a new SQL query for every combination of token+recipient. Batch-loading is better.

But I've got this nagging feeling that it would be better to significantly trim down `alterActionScheduleQuery()`:
* _Change_: For `alterActionScheduleQuery()`,  we don't need a blanket `SELECT e.*` or any option-group joins.
* _Change_: For `prefetch()`, get the list of activity IDs from a mix of (a) `context['activityId']`and/or (b) `context['actionSearchResult']->entity_id` (*provided the mapping type is right*).
* _Observe_: It always loads activity data the same way.
* _Observe_: `prefetch()` only runs once for a given batch, so we're still batched-loading.
* _Observe_: When running the scheduled-reminders, the main query can be a bit thinner (easier to read/debug).
* _Observe_: If we optimize `prefetch()` a bit more (*only grab columns that are actually needed*), then that optimization applies equally well to both use-cases.

Reply:
Thanks for your thoughtful comments. 

I haven't touched `alterActionScheduleQuery()` or looked at that process in much detail as it was pre-existing and only used for scheduled reminders. The concept of a base SQL query that is passed around for various subscribers to modify for their own needs seems clever, and probably performant, but easy to mess up. 

As you point out the schemas for `Activity.get` API and `civicrm_activity` may diverge so tracking both could get confusing.  Which one do you prefer for these purposes?

Reply:
@eileenmcnaughton Thanks for the steer - I'll go that way.

Reply:
@eileenmcnaughton @totten I set off down the path of following `civicrm_activity` rather than `Activity.get`.  Then I got to the comments on `CRM_BAO_Activity::retrieve()` noting it should be deprecated and to use the API instead.  if we go direct to the table or via the BAO we hit the same issue of bypassing ACL's, so I've stuck with the API, at least for now.

However, by explicitly specifying the return fields, the available tokens are the ones specified in `__construct()` rather than whatever the API happens to return.

I've made the changes Tim suggested to trim down `alterActionScheduleQuery()` and move the data gathering to `prefetch()` so it is consistent between scheduled reminders and other uses.